mkosi/machine.py

```diff
@@ -0,0 +1,102 @@
+import signal
+import subprocess
+from typing import TYPE_CHECKING, Any, Optional, Sequence
+
+import pexpect  # type: ignore
+
+from . import (
+    MKOSI_COMMANDS_SUDO,
+    build_stuff,
+    check_native,
+    check_output,
+    check_root,
+    init_namespace,
+    load_args,
+    needs_build,
+    parse_args,
+    prepend_to_environ_path,
+    run_command_image,
+    run_qemu_cmdline,
+    run_shell_cmdline,
+    unlink_output,
+)
+from .backend import die
+
+if TYPE_CHECKING:
+    CompletedProcess = subprocess.CompletedProcess[Any]
+else:
+    CompletedProcess = subprocess.CompletedProcess
+
+
+class Machine:
+    def __init__(self, args: Optional[Sequence[str]] = None, debug: Optional[bool] = False) -> None:
+        # Remains None until image is built and booted, then receives pexpect process
+        self._serial: Optional[pexpect.spawn] = None
+        self.stack = contextlib.ExitStack()
+        self.args = load_args(parse_args(args)["default"])
+        self.exit_code: int = 1
+        self.debug = debug
+
+
+    @property
+    def serial(self) -> pexpect.spawn:
+        if self._serial is None:
+            raise ValueError("Trying to access serial console before machine boot or after machine shutdown.")
+
+        return self._serial
+
+
+    def __enter__(self) -> Machine:
+        if self.args.verb in MKOSI_COMMANDS_SUDO:
+            check_root()
+            unlink_output(self.args)
+
+        if self.args.verb == "build":
+            check_output(self.args)
+
+        if needs_build(self.args):
+            check_root()
+            check_native(self.args)
+            init_namespace(self.args)
+            build_stuff(self.args)
+
+        with contextlib.ExitStack() as stack:
+            prepend_to_environ_path(self.args.extra_search_paths)
+
+            if self.args.verb in ("shell", "boot"):
+                cmdline = run_shell_cmdline(self.args)
+            elif self.args.verb == "qemu":
+                # We must keep the temporary file opened at run_qemu_cmdline accessible, hence the context stack.
+                cmdline = stack.enter_context(run_qemu_cmdline(self.args))
+            else:
+                die("No valid verb was entered.")
+
+            cmdline = [str(x) for x in cmdline]
+            cmdline = " ".join(cmdline)
+
+            self._serial = pexpect.spawnu(cmdline, logfile=None, timeout=240)
+            self._serial.expect("#")
+            self.stack = stack.pop_all()
+
+        return self
+
+
+    def run(self, commands: Sequence[str]) -> CompletedProcess:
+        process = run_command_image(self.args, commands, subprocess.PIPE, subprocess.PIPE)
+        if self.debug:
+            print("Stdout:\n" + process.stdout)
+            print("Stderr:\n" + process.stderr)
+
+        return process
+
+
+    def __exit__(self, *args: Any, **kwargs: Any) -> None:
+        if self._serial is not None:
+            self._serial.kill(signal.SIGTERM)
+            self.exit_code = self._serial.wait()
```

Comment:
We should assign `None` again to `self._serial` after we kill the process so it can't be used after the machine is shut down.